wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/tool/kubernetes/ContainerSearch.java

```diff
@@ -0,0 +1,166 @@
+import io.fabric8.kubernetes.api.model.KubernetesList;
+import io.fabric8.kubernetes.api.model.KubernetesResource;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.api.model.PodTemplateSpec;
+import io.fabric8.kubernetes.api.model.apps.Deployment;
+import io.fabric8.openshift.api.model.Template;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import org.eclipse.che.commons.annotation.Nullable;
+
+/**
+ * Container search goes through a list of Kubernetes resources and recursively looks for containers
+ * that match the provided criteria.
+ *
+ * <p>The deployment and pod name criteria work both on the {@code name} and, if name is not set on
+ * given deployment/pod, on the {@code generateName} of the objects.
+ */
+public class ContainerSearch {
+
+  private final @Nullable String deploymentName;
+  private final @Nullable String podName;
+  private final @Nullable String containerName;
+  private final @Nullable Map<String, String> podSelector;
+  private final @Nullable Map<String, String> deploymentSelector;
+
+  /**
+   * Constructs a new {@code ContainerSearch} instance in somewhat unsurprising manner.
+   *
+   * @param deploymentName only search for containers in pods of deployments with given name
+   * @param podName only search for containers in pods with given name
+   * @param containerName only search for containers with given name
+   * @param podSelector only search for containers in pods with labels matching the selector
+   * @param deploymentSelector only search for containers in pods of deployments with labels
+   *     matching the selector
+   */
+  public ContainerSearch(
+      @Nullable String deploymentName,
+      @Nullable String podName,
+      @Nullable String containerName,
+      @Nullable Map<String, String> podSelector,
+      @Nullable Map<String, String> deploymentSelector) {
+    this.deploymentName = deploymentName;
+    this.podName = podName;
+    this.containerName = containerName;
+    this.podSelector = podSelector;
+    this.deploymentSelector = deploymentSelector;
+  }
+
+  /**
+   * Searches for containers in the provided list of Kubernetes resources. If any given item in the
+   * list can contain a container (i.e. it is a pod, deployment, Kubernetes list or Openshift
+   * template) the item is searched for the containers recursively.
+   *
+   * @param list the list of Kubernetes resources to sift through
+   * @return a list of containers found in the provided resource list
+   */
+  public List<Container> search(Collection<? extends KubernetesResource> list) {
+    List<Container> ret = new ArrayList<>();
+
+    search(list, ret, false, false);
+
+    return ret;
+  }
+
+  private void search(
+      Collection<? extends KubernetesResource> list,
+      Collection<Container> results,
+      boolean insideDeployment,
```

Comment:
I think the usage of `insideDeployment` and `insidePod` make code harder to understand.
What do you think about such an alternative? 
https://github.com/sleshchenko/che/commit/4bc2504db0f234cfa2f2b5c8fab0d9efb46ab77f

All unit tests passed but I can not guarantee that all use cases you kept in mind works fine.